x/clp/keeper/rewards.go

```diff
@@ -0,0 +1,258 @@
+package keeper
+
+import (
+	"strings"
+
+	"github.com/Sifchain/sifnode/x/clp/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	abci "github.com/tendermint/tendermint/abci/types"
+)
+
+// BeginBlock prunes / auto-cancels the providers expired unlock requests.
+func (k Keeper) BeginBlock(ctx sdk.Context) {
+	params := k.GetParams(ctx)
+	for it := k.GetLiquidityProviderIterator(ctx); it.Valid(); it.Next() {
+		value := it.Value()
+		var lp types.LiquidityProvider
+		if len(value) <= 0 {
+			continue
+		}
+		err := k.cdc.Unmarshal(value, &lp)
+		if err != nil {
+			continue
+		}
+		k.PruneUnlockRecords(ctx, lp, params.LiquidityRemovalLockPeriod, params.LiquidityRemovalCancelPeriod)
```

Comment:
Can we move this function call to `Remove Liquidity` and `Swap`.The logic being 
- User requests remove liquidity/swap 
- Calculate lpuints
- Expire Unlocks 
- Check if user has enough lpuints 

The double for loop at begin block might be quite heavy , considering the number of lp's, Since an LP is unique to address + asset

Reply:
That's a good point. The advantage of doing it here is that we can trigger an event with the right cancellation block height. I think we can probably relax that requirement esp because UI is not showing a history from data services anymore, but we should chat to the team to let them know.

Reply:
In that case UseUnlockedLiquidity should also ensure the record is not passed a cancel period.

Reply:
Yeah I can see how iterating through every LP could be heavy and possibly an attack vector to slow down the network even.

Triggering it on other address actions seems like a good workaround to me.